Add bilateralization tool (project/newpathways-trade)

junukitashepard opened the pull request:
## Overview
The merged code allows users to represent trade bilaterally for specified commodities. By default, the tool bilateralizes existing fuel trade (e.g., oil_exp).

The bilateralization tool, `bilateralize`, is generalized for any traded commodity, whether that is a fuel (e.g., LNG), or a material (e.g., steel). It also explicitly represents bilateral trade “flows”, or how a fuel/commodity is transported from exporter to importer. These flow technologies are user defined and flexible; the most common are pipelines (e.g., gas pipelines), maritime shipping (e.g., LNG tanker), and transmission lines. 

See [https://github.com/iiasa/message-ix-models/blob/project/newpathways-trade/doc/api/tools-bilateralize.rst](url) for more details on implementation.

## How to review

Note: All scripts are additions (no existing scripts are amended). Therefore, I would request reviewing code directly rather than individual commits (of which there are >200).

These review steps are also outlined in [https://github.com/iiasa/message-ix-models/blob/project/newpathways-trade/message_ix_models/tools/bilateralize/workflow.py](url)
1. The function `prepare_edit_files()` should produce untracked files in commodity-specific directories in `data/bilateralize/edit_files` and also transfer required parameter files to `data/bilateralize/bare_files`
2. The function `bare_to_scenario()` should produce a dictionary of required parameter updates for bilateralization for all specified trade commodities in `config.yaml`.
3. The function `load_and_solve()` should be able to be used to update SSP base scenarios
4. The documentation should make sense even though it is incomplete. It is currently conceptual and still needs function documentation.

<!--
For example, one or more of:

- Read the diff and note that the CI checks all pass.
- Run a specific code snippet or command and check the output.
- Look at a certain page in the ReadTheDocs preview build of the documentation.
- Ensure that changes/additions are self-documenting, i.e. that another
  developer (someone like the reviewer) will be able to understand what the code
  does in the future.
-->

## PR checklist

<!-- This item is always required. -->
- [ ] Continuous integration checks all ✅
  <!--
  The following items are all *required* if the PR results in changes to user-
  facing behaviour, e.g. new features or fixes to existing behaviour. They are
  *optional* if the changes are solely to documentation, CI configuration, etc.

  In ambiguous cases, strike them out and add a short explanation, e.g.

  - ~Add or expand tests.~ No change in behaviour, simply refactoring.
  -->
- [x] Add or expand tests; coverage checks both ✅
- [x] Add, expand, or update documentation.
- [x] Update doc/whatsnew.
  <!--
  To do this, add a single line at the TOP of the “Next release” section of
  doc/whatsnew.rst, where '999' is the GitHub pull request number:

  - Title or single-sentence description from above (:pull:`999`:).

  Commit with a message like “Add #999 to doc/whatsnew”
  -->

junukitashepard committed:
Initial bilateralize tool and config

junukitashepard committed:
Build and test bare sheets for oil_shipped

junukitashepard committed:
Update .gitignore

Add lines s.t. repo ignores bare, edit, and diagnostic files

junukitashepard committed:
Simplify gas pipeline framework

junukitashepard committed:
Add diagnostics and calibration

junukitashepard committed:
Remove scenario run requirement

junukitashepard committed:
Add historical calibration

junukitashepard committed:
Add historical calibration

junukitashepard committed:
Add tracked metrics to config

junukitashepard committed:
Add sankey for diagnostics of bilateral trade

junukitashepard committed:
Link to GEM for historical infrastructure data

junukitashepard committed:
Add historical calibration and option to output gdx (no solve)

junukitashepard committed:
Add flow technologies to full function (bilateral only)

Tested for pipelines only

junukitashepard committed:
Run with steel data inputs

junukitashepard committed:
Add flow tec to diagnostics

junukitashepard committed:
Update sankey.py

junukitashepard committed:
Add flows and proxy for 2025 with 2023 activity

junukitashepard committed:
Specify gas pipes in historical calibration

junukitashepard committed:
Make sure base technology gets deleted

junukitashepard committed:
Final global flow constraint commit

junukitashepard committed:
Add MariTEAM and calibrate LNG

junukitashepard committed:
Restructure code between /tools and /project

junukitashepard committed:
Clean up pipeline calibration and GEM connection

junukitashepard committed:
LFS update to LNG Tankers

junukitashepard committed:
Technical lifetime for LNG to 20 years

junukitashepard committed:
Run diagnostics for LNG and pipeline

junukitashepard committed:
Create newpathways-trade.rst

This will get replaced with bilateralization documentation later

junukitashepard committed:
Set up generate_bare_sheets so default values populated in bare_files

junukitashepard committed:
Refactor and run bilateralize.py

bilateralize.py now relies on the message_ix.make_df function rather than a user-defined function. Cuts down code ~120 lines.

junukitashepard committed:
Update flow formulation in bilateralization

In updated framework, flow technologies produce commodities (capacity) that are inputs to trade technologies, rather than using relation_activity to link flow-trade.

junukitashepard committed:
Add historical_new_capacity for pipelines

junukitashepard committed:
Add year_vtg and begin at 2000 for historical_new_capacity of pipelines

junukitashepard committed:
Rerun edit files with LNG flow tech as input to trade tech

junukitashepard committed:
Remove gas_piped from .gitignore

junukitashepard committed:
Update config_default.yaml

junukitashepard committed:
Add p-drive flexibility to GEM and MariTeam call

P-drive flexibility in pull_gem.py and mariteam_calibration.py

junukitashepard committed:
Add project name/config to bare_to_scenario

junukitashepard committed:
Rebuild BACI for historical_calibration

junukitashepard committed:
Set up load_and_solve

bare_to_scenario will not actually touch scenario in ixmp

junukitashepard committed:
Update load_and_solve

make sure config_path is pointing

junukitashepard committed:
Set up LNG with updated flow framework

LNG tanker capacity as inputs to trade

junukitashepard committed:
Set config to solve MESSAGE

junukitashepard committed:
Correct foil commodity to fueloil

junukitashepard committed:
Add cost to LNG trade

junukitashepard committed:
Update variable cost for LNG trade

junukitashepard committed:
Remove pydeflate

junukitashepard committed:
Add variable cost for LNG shipping

junukitashepard committed:
Move variable cost to investment cost for LNG shipping

junukitashepard committed:
Add fix cost to LNG shipping

junukitashepard committed:
Add activity constraints to LNG shipping trade technology

junukitashepard committed:
Remove constraint extend tec lifetime for LNG trade tec

junukitashepard committed:
Add growth and initial activity constraints to LNG trade tec

junukitashepard committed:
Add Sankey and Chord functionalities

junukitashepard committed:
Differentiate distances for maritime shipping

junukitashepard committed:
Add LNG tanker historical new capacity, lower LNG trade growth constraints

junukitashepard committed:
Change year_vtg to integer in historical new capacity

junukitashepard committed:
Update config.yaml

P-drive location

junukitashepard committed:
Add trade historical new capacities (not just flow)

junukitashepard committed:
Make year_vtg integer in hist_new_capacity for trade

junukitashepard committed:
Add gas pipeline to RCPA

junukitashepard committed:
Add soft activity bounds to LNG trade tec

junukitashepard committed:
Update var_costs and tec_lt for LNG tankers

Assume $3/mcf for tankers, 5 years for trade decisionmaking

junukitashepard committed:
Add cost scenarios

Re-add variable cost as default

junukitashepard committed:
Add level_cost_activity_soft and abs_cost_activity_soft constraints for LNG, add additional parameter update option to clone_and_update

junukitashepard committed:
Set var_cost to 5 year for tankers, add scenario for NAM shale discount

junukitashepard committed:
Update inv_cost based on higher lifetime Mt-km

junukitashepard committed:
Rerun all scenarios

junukitashepard committed:
Remove technical lifetimes from config.yaml

junukitashepard committed:
Reset broadcasting function in bilateralization

junukitashepard committed:
Add var cost to LNG trade

junukitashepard committed:
Re-add all test scenarios

No flow costs on LNG shipped (LNG_noFLcost), no trade fix costs on LNG shipped (LNG_noTRfixcost), reduce NAM shale variable costs to MEA (lowNAMshalecost)

junukitashepard committed:
Adjust variable costs on LNG

junukitashepard committed:
Update load_and_solve.py

junukitashepard committed:
Add options for flow capacity constraints; debug trade_parameter dictionary

trade_parameter dictionaries: references were being updated instead of copies, instead just re-import parameter dictionary for each test scenario.

junukitashepard committed:
Add distance script to generalize further regionalization

junukitashepard committed:
Update test scenarios

Comment out default
Add no variable trade costs scenario
Add high LNG tanker fuel scenario
Make all gas extraction cheaper in low NAM cost scenario

junukitashepard committed:
Add var_cost to bare by default, add more test scenarios

junukitashepard committed:
Fix pathname in pull_gem

MESSAGE_Trade to MESSAGE_trade

junukitashepard committed:
Fix pathname in mariteam

MESSAGE_Trade to MESSAGE_trade

junukitashepard committed:
Remove max/min on var costs

trade variable costs

junukitashepard committed:
Adjust test scenario on var cost

Instead of lowering var cost, increasing var cost

junukitashepard committed:
Set default variable cost to mean instead of min

junukitashepard committed:
Round var cost values

junukitashepard committed:
Comment out test scenarios that were already run

junukitashepard committed:
Add up-to-date test scenarios in diagnostics

junukitashepard committed:
Add steel back; add regionality to bilateral commodities

Gas pipeline commodity should be gas_pipeline_capacity_[NODE]

junukitashepard committed:
Set up liquefaction spreadsheet; only run baseline test scenario (pipelines_LNG)

junukitashepard committed:
Add liquefaction test scenario

junukitashepard committed:
Debug: trade_parameters add

junukitashepard committed:
Add base diagnostics and clean up newpathways project repo

junukitashepard committed:
Update Russian LNG terminal from Novorossiysk to Arctic LNG 2

junukitashepard committed:
Transfer Russina LNG terminal update to bare file

junukitashepard committed:
Update scenario pickle

Steel input update and Russia LNG terminal update

Repoint for UnICC

junukitashepard committed:
Remove steel as input to activity

Instead draw based on CAP_NEW

junukitashepard committed:
Add HHI test scenarios

junukitashepard committed:
Remove reporting tables

junukitashepard committed:
Use path join for hhi_config name

junukitashepard committed:
Set HHI test to WEU only (assume CHN will break)

junukitashepard committed:
Small debug

junukitashepard committed:
Add coal representation

junukitashepard committed:
Finish adding coal, option to remove domestic_coal/gas relation activity

domestic_coal is used to constrain PAO primary energy consumption.

junukitashepard committed:
Set up constraints on coal

junukitashepard committed:
Add shipped crude

junukitashepard committed:
Add crude oil pipelines

junukitashepard committed:
Update GEM calibration for oil (and gas) pipelines

Build long dataframe up front so that historical activity can be recorded in 5 year increments, instead of just for 2025

junukitashepard committed:
Set load_and_solve to solve ixmp

junukitashepard committed:
Ensure years are integers in historical_activity

junukitashepard committed:
Remove var cost on crude oil shipping

junukitashepard committed:
Add ethanol shipping

junukitashepard committed:
Add 2025 values

junukitashepard committed:
Update costs to shipped and piped crude

junukitashepard committed:
Add fuel oil trade

junukitashepard committed:
Add costs to fuel oil shipping

junukitashepard committed:
Add fuel oil piped trade

junukitashepard committed:
Add light oil piped and shipped

junukitashepard committed:
Ensure that years in GEM based files are integers

MESSAGE does not read year_act as float

junukitashepard committed:
Light oil should be used for shipping by default

Not fuel oil

junukitashepard committed:
LNG shipping updated to fueled by light oil

junukitashepard committed:
Remove piped light oil for debugging

junukitashepard committed:
Set load_and_solve to solve (not to_gdx)

junukitashepard committed:
Remove NaN from output for flow_tech loil

junukitashepard committed:
Add 1 to outputs for shipped coal flow

junukitashepard committed:
Add loil_piped

junukitashepard committed:
Add biomass shipping

junukitashepard committed:
Add lh2 trade

junukitashepard committed:
Rename H2 scenario and updated diagnostics

junukitashepard committed:
Set base trade tecnames so that base technologies are removed

For example, crudeoil_shipped_exp has a base trade technology of oil_exp.

junukitashepard committed:
Add fuel oil bunker technology to foil_ship

This is to connect to foil_bunker technology, needed for shipping workflow

junukitashepard committed:
Add cost and growth constraints for sensitivities

junukitashepard committed:
Remove constraints for fuel oil, lh2, loil

Testing

junukitashepard committed:
Add pickle back to load_and_solve.py

Also run test scenarios on UniCC

junukitashepard committed:
Debug LH2 configuration

Configuration for LH2 was using biomass historical activity

junukitashepard committed:
Test scenario after LH2 debug

junukitashepard committed:
Set constraint parameter removal to be flexible

Constraint parameters only removed if they exist in dictionary

junukitashepard committed:
Update .gitignore to ignore diagnostic data

Gitignore scenario_parameters.pkl

junukitashepard committed:
Remove caches for untracked files

Not tracking bare, edit, GEM infrastructure, and diagnostic files

junukitashepard committed:
Move configs to technology-specific yamls

junukitashepard committed:
Delete build_sankey.py

junukitashepard committed:
Update configs

Separate all configs, add trade units (so both energy and materials can be bilateralized)

junukitashepard committed:
Remove inter_pipe (artifact)

junukitashepard committed:
Add workflow and test

junukitashepard committed:
Allow user specification on scenario updates (rather than config)

junukitashepard committed:
Allow diagnostics for NEWPATHWAYS for all commodities

junukitashepard committed:
Don't allow pipelines from FSU to PAO

junukitashepard committed:
Add information on gas compression along pipelines

junukitashepard committed:
Add scenario flexibility to load_and_solve

junukitashepard committed:
Ensure historical activity loads only if pipeline is specified

junukitashepard committed:
Refactor bilateralize.py

junukitashepard committed:
Remove project/newpathways from branch

put in separate project/newpathways_project branch

junukitashepard committed:
Remove implied optionals

junukitashepard committed:
Split bilateralize into step code

1) prepare_edit to generate edit files
2) bare_to_scenario to generate dictionary
3) load_and_solve to update scenario and optionally solve

junukitashepard committed:
Reformat to meet mypy, ruff, and testing

Update formatting to pass mypy/ruff and allow testing

junukitashepard committed:
Create documentation tools-bilateralize

This removes the NEWPATHWAYS project version of documentation (same content)

junukitashepard commented on message_ix_models/tests/tools/test_bilateralize.py:
```diff
@@ -0,0 +1,129 @@
+
+
+# Test generate_bare_sheets()
+@pytest.fixture
```
Thanks Paul, I'll review this now.

junukitashepard committed:
Remove pytest fixtures from test_bilateralize.py

junukitashepard committed:
Update test_bilateralize and associated files

Load and solve is now tested
Debug to load_and_solve.py (add "_")
Add encoding to prepare_edit.py

junukitashepard committed:
Add to #438 to doc/whatsnew.rst

junukitashepard committed:
Add more tests to test_bilateralize

Increase coverage to GEM and MariTeam imports

junukitashepard committed:
Reformat prepare_edit for mypy and ruff

junukitashepard committed:
Update test_bilateralize.py to meet mypy

junukitashepard committed:
Satisfy ruff in test_bilateralize

junukitashepard commented:
@khaeru Patch coverage is 51%; however, the scripts that are not covered in testing are those that derive input data based on private data sources. I would of course welcome any input on whether/how to increase this coverage.

junukitashepard committed:
Update output directory in pull_gem.py

Should be in separate GEM folder

junukitashepard committed:
Point project and config names in subfunctions

khaeru commented on message_ix_models/data/bilateralize/node_lists/R12_node_list.yaml:
```diff
@@ -0,0 +1,121 @@
+# Region code list
```
Are these files any different from files like data/node/R12.yaml?

If they aren't, we should remove the duplicates and use existing code that loads and manipulates the existing files.

If they are, please explain how exactly. I would guess we can find ways to incorporate any information needed by .tools.bilateralize into the existing files.

khaeru committed:
Add tests of .bilat.{historical_calibration,pull_gem}

khaeru commented on doc/api/tools-bilateralize.rst:
```diff
@@ -0,0 +1,97 @@
+.. currentmodule:: mesage_ix_models.tools.bilateralize
```
This added file will not appear in the built documentation until it is listed in a 'toctree'. The places to insert it are here:
https://github.com/iiasa/message-ix-models/blob/2bc42a33b1a8574afcb785e772feb9fc0ddda560/doc/index.rst?plain=1#L102-L104
and here:
https://github.com/iiasa/message-ix-models/blob/2bc42a33b1a8574afcb785e772feb9fc0ddda560/doc/api/tools.rst?plain=1#L9-L11

…following the style of the existing entries, and in alphabetical order.

junukitashepard replied:
Added in https://github.com/iiasa/message-ix-models/pull/438/commits/7d2157961d7d5a78ac1b16a6797e9d6abb75e518

junukitashepard committed:
Remove data/bilateralize/node_lists and references

junukitashepard commented:
TODO: Update data references to IEA:
* WEB should use tools.iea.web.load_data() with update: https://github.com/iiasa/message-ix-models/blob/2bc42a33b1a8574afcb785e772feb9fc0ddda560/message_ix_models/tools/iea/web.py#L82-L92
* NGD code needs to be added to tools.iea
* CEPII has its own issue and will continue to refer to P-drive until tool is added https://github.com/iiasa/message-ix-models/issues/449

junukitashepard committed:
Add opiton to load and solve scenario directly

Set scenario to a message_ix.Scenario for direct bialteralization in load_and_clone
Otherwise function uses start model/scenario from config.yaml
There is also option to name specific start model/scenario

junukitashepard committed:
Create __init__.py

junukitashepard committed:
Add bilateralize documentation to doc indices

khaeru commented:
I see that with recent commits [the new docs page](https://iiasa-energy-program-message-ix--438.com.readthedocs.build/projects/models/en/438/api/tools-bilateralize.html) now appears in the preview build! 🥳

junukitashepard committed:
Update file locations in pull_gem.py

junukitashepard committed:
Only remove base oil trade if crude oil is called

To remove oil_exp and oil_imp

junukitashepard committed:
Change GEM to csv in P-drive

junukitashepard committed:
Add relation activity for regional exports

Necessary for legacy reporting update

junukitashepard committed:
Ensure project/config name specified in build_historical_price

junukitashepard committed:
Update testing and formatting to meet ruff/mypy

junukitashepard committed:
Formatting for ruff

junukitashepard committed:
Change var_cost to fix_cost for trade parameters

yiyi1991 commented:
Hi Jun, here are some points from mine. 

- Do we already or can we open an issue for the bilateralization? Tracking how at the end it may look like, to-do list, what is covered by which PR, what is going to be covered by which PR, etc. You **then do not have to tackle all the points listed below** in this PR. Feel free to take some to the issue list and process later. 

- The design of the template config is comprehensive! I like it a lot.  

- **[Term clarification]** The following terms need to be better defined in the documentation: what it is, it is an accounting technology or a real technology with a lifetime, general examples, and the corresponding technology name in the latest R12 scenario. 
  - trade technology
  - flow technology
  - and two other categories, supply technology and bunker technology (the naming of these two is not as self-explanatory as trade tech and flow tech, but fare enough)
- **[De-project-ization]** In general we should avoid describing a work by project names. I also realized that every time you call `project_name`, you do not mean a specific combination of bilateralized commodities (A+B+C commodities for project X, D commodity for project Y), but simply the path that users are working on. If my understanding is correct, can we just use `path_name` instead?

- **[Workflow]** I myself prefer a separate build, solve, and report. Binding build and solve together and providing another option to save gdx also makes sense. But the name of the function `load_and_solve()` is too confusing. Will you consider `build`, or at least `build_and_solve`? Another confusing one is `bare_to_scenario()`. In the definition of the functions, it will also be useful to clarify what will be returned. Key users want to know which step a Scenario object is going to be taken in. 
```
prepare_edits() -> returns editable csv files
bare_to_scenario() -> returns a dictionary (does not return a scenario!)
load_and_solve() -> returns a (solved) scenario
```
- **[Workflow refactor]** Are you familiar with the current `make_spec()` (require/add/remove), `apply_spec()`, and `add_par_data()` workflow used in build materials/water/transportation/cost tool? You may find that some of your `add_trade_data()` can apply `add_par_data()`. I do not insist on a refactor. As I was able to build bilateralization of all commodities covered by the current default config on a recent R12 scenario in **14** minutes on my machine. Nothing to complain about the performance.  

- **[Generate commodity/technology list]** It will be great if new trade and flow technologies, as well as new trade commodities, are generated in the commodity.yaml or technology.yaml too. Maybe you are already working on this in the report PR. This autogeneration is also one benefit of `make_spec()` -> `apply_spec()` -> `add_par_data()` style. 
- **[Tutorial]** One showcase of how users can use it (e.g., higher cost of a specific trade pair on a specific shipped commodity) so they understand what they do NOT need to touch in a default setup. But the whole tutorial part can be done later. 

After code quality reaches the target test coverage and the UnicodeDecodeError on Windows is tackled properly, I would happily approve this merge.

junukitashepard committed:
Allow for user to set base trade technologies

khaeru commented on message_ix_models/data/bilateralize/desktop.ini:
OS-specific files like this should not be committed. To hide them from Git, you could add them to the top-level .gitignore file, e.g. here is a file that I believe is macOS-specific: https://github.com/iiasa/message-ix-models/blob/6bf400012bc5d343cc0e9d01cc1ee5a3fde73d04/.gitignore#L149

khaeru commented on doc/api/tools-bilateralize.rst:
```diff
@@ -0,0 +1,97 @@
+  
+Overview
+========
+This documentation will outline how to bilateralize trade flows in MESSAGEix. By default, trade flows are based on a “global pool” approach, wherein all exporters export energy/commodities to a pool from which importers import based on regionally resolved demands.
```
Please see the [documentation code style](https://docs.messageix.org/en/latest/contributing.html#documentation) per "semantic line breaks".

As an intermediate step, start at least every sentence on a new line.

junukitashepard committed:
Update utf encoding when reading yaml

khaeru commented on doc/api/tools-bilateralize.rst:
```diff
@@ -0,0 +1,97 @@
+=============
+The first step is to generate empty (or default valued) parameters that are required for bilateralization, specified by commodity. This step requires updates to a configuration file (``config.yaml``) that should be housed in a project directory (e.g., ``message-ix-models/projects/newpathways-trade/config.yaml``). A template configuration file is provided at ``message-ix-models/data/bilateralize/c ...
+
+Once the configuration is updated, the user can run ``message_ix_models.tools.bilateralize.prepare_edit.generate_edit_files(log, project_name, config_name, message_regions)`` to produce empty (or default valued) parameters as CSV files. These CSV files will populate in ``message-ix-models/data/[your_trade_commodity]/edit_files``. 
```
Please see the existing documentation and the Sphinx docs ([1](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html), [2](https://www.sphinx-doc.org/en/master/usage/referencing.html), [3](https://www.sphinx-doc.org/en/master/usage/domains/python.html#cross-referencing-python-objects)) for how to use 'roles'.

Some that could be used in this file include:
```rest
:file:`message_ix_models/data/bilateralize/configs/base_config.yaml`
:func:`.bilateralize.prepare_edit.generate_edit_files`
:mod:`.bilateralize.workflow`
```

junukitashepard committed:
Add technical lifetime as a user configuration

junukitashepard committed:
Use existing util.broadcast in bare_to_scenario

junukitashepard committed:
Update tools-bilateralize.rst

junukitashepard committed:
Update tools-bilateralize.rst

junukitashepard committed:
Debug bilateralize test

junukitashepard committed:
Add flexibility to bare_to_scenario

junukitashepard committed:
Ensure shipped costs transfer from edit to bare

junukitashepard committed:
Add test back

junukitashepard committed:
Updates based on alps workflow

junukitashepard committed:
Ruff formatting

junukitashepard committed:
Add test for bare_scenario

junukitashepard committed:
Add test coverage

junukitashepard committed:
Update test_bare_to_scenario.py

junukitashepard committed:
Set test_calibrate_historical_shipping as xfail

junukitashepard committed:
Reformat for ruff

junukitashepard committed:
Add scgraph to pyproject.toml

For bilateralize tool and tests

junukitashepard committed:
Add test coverage for mariteam_calibration

yiyi1991 commented:
Points in #460 are basically covered. (Maybe we try refactoring to make_spec() -> apply_spec() -> add_par_data() later)
Happy to see that it lowered the cognitive load for potential users and strengthened the reliability with tests. There are always things that can be improved but maybe kick this off first as a base.

khaeru committed:
Improve doc/tools-bilateralize.rst

- Correct typo in .. currentmodule:: directive.
- Use recommended section symbol hierarchy.
- Extend section markup to match length of section title line.
- Correct indentation, blank lines for bulleted lists.
- Add recursive autosummary for the module code.

khaeru reviewed (approved):
Approved based on:
- Passing checks, notwithstanding codecov/patch as mentioned [here](https://github.com/iiasa/message-ix-models/pull/438#issuecomment-3823566717).
- Comments by @yiyi1991, which I understand are captured at #460, #462 for follow-up in the next PR.
- Recent rebase and clean-up. In doing this I opened #476 for docs-related comments from November that still need to be addressed.

NB I have _not_ reviewed the code in detail, as the number of commits and amount of added code is too large. But now that the module exists, we can be more particular about its contents as they are improved in future PRs.

I'll merge as soon as the checks pass.
